### PR TITLE
Emsystem

### PR DESCRIPTION
**Pull request by tiffanyhsyu**

### CODE REVIEW

File: linetools/analysis/emline.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      Name of the method to be applied
+      PG16 --
+          Requires Hbeta, [OII], [OIII], [NII], [SII]
+          Return r_val and s_val
+    emsystem : EmSystem
+
+    Returns
+    -------
+    """
+    if method == 'PG16':
```
Comment (profxj) [inline linetools/analysis/emline.py]: Can you add a http link to the paper here?

File: linetools/isgm/emsystem.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+        # Return
+        return slf
+
+    @classmethod
+    def from_alis(cls, alis_file, radec, **kwargs):
+        """ Load from an ALIS output file
+
+        Parameters
+        ----------
+        alis_file
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: add

``` : str
```

File: linetools/isgm/emsystem.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+        """
+        idict = ltu.loadjson(json_file)
+        slf = cls.from_dict(idict, **kwargs)
+        return slf
+
+
+    def __init__(self, radec, zem, vlim=None, em_type=None, name=None):
+
+        self.zem = zem
+        if vlim is None:
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: We need to use the zLimits Class here.

I can help set that up.  Remind me.

File: linetools/isgm/emsystem.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+            if not testcoord:
+                warnings.warn('Failed coordinate match')
+            if not testcomp:
+                warnings.warn('Failed component check')
+            if not testz:
+                warnings.warn('Failed velocity overlap')
+        #
+        return test
+
+    def chk_emline(self, component):
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: We don't use components with EmLine (at least not now).

Remove components and/or this method.  

Same goes for anywhere else you see components.

File: linetools/isgm/emsystem.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+        """
+        return True
+
+    def fill_ionN(self, **kwargs):
+        """ Fills the ionN Table from the list of components
+        """
+        self._ionN = ltiu.iontable_from_components(self._components, **kwargs)
+
+    def fill_trans(self, **kwargs):
+        """ Fills the ionN Table from the list of components
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: list of emlines

File: linetools/isgm/emsystem.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+        """
+        self._trans = ltlu.transtable_from_speclines(self.list_of_abslines())
+
+    def get_emline(self, inp):
+        """ Returns an EmLine from the AbsSystem
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: EmSystem

File: linetools/isgm/emsystem.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+
+        Parameters
+        ----------
+        inp : str or Quantity
+          str -- Name of the transition, e.g. 'CII 1334'
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: Give an EmLine example, e.g. Halpha

File: linetools/isgm/emsystem.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+        # Finish
+        if len(mt) == 0:
+            warnings.warn("No emline with input={}".format(inp))
+            return None
+        elif len(mt) == 1:
+            return self._emlines[mt[0]]
+        else:
+            return [self._emlines[ii] for ii in mt]
+
+    def get_component(self, inp):
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: remove this method

File: linetools/isgm/emsystem.py [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+            except ValueError:
+                pass
+            else:
+                return comp
+        # Raise error?
+        warnings.warn("Input absorption line is not in any component")
+        return None
+
+    def measure_restew(self, spec=None, **kwargs):
+        """ Measure rest-frame EWs for lines in the AbsSystem
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: EmSystem

File: linetools/isgm/emsystem.py [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+          Over-ride system vlim for plotting
+        """
+        from linetools.analysis import plots as ltap
+        if pvlim is not None:
+            vlim = pvlim
+        else:
+            vlim = self.vlim
+        ltap.stack_plot(self.list_of_abslines(), vlim=vlim, **kwargs)
+
+    def to_dict(self):
```
Comment (profxj) [inline linetools/isgm/emsystem.py]: Does this work?

File: linetools/isgm/tests/test_init_emsys.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+from astropy.coordinates import SkyCoord
+
+from linetools.isgm.emsystem import EmSystem
+
+
+def data_path(filename):
+    data_dir = os.path.join(os.path.dirname(__file__), 'files')
+    return os.path.join(data_dir, filename)
+
+
```
Comment (profxj) [inline linetools/isgm/tests/test_init_emsys.py]: Add some other, simpler tests for instantiating the object.

File: linetools/tests/test_io.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import linetools.io as lio
+
+
+def data_path(filename):
+    data_dir = os.path.join(os.path.dirname(__file__), 'files')
+    return os.path.join(data_dir, filename)
+
+def test_emline_from_alis():
+    lio.emlines_from_alis_output(data_path('spec1d_J0018p2345_KASTb_coadd.mod.out'))
+
```
Comment (profxj) [inline linetools/tests/test_io.py]: Add some checks that this actually worked.

File: linetools/analysis/emline.py (hunk 1 above)
Comment (tiffanyhsyu) [reply linetools/analysis/emline.py]: This has been added

File: linetools/isgm/emsystem.py (hunk 3 above)
Comment (tiffanyhsyu) [reply linetools/isgm/emsystem.py]: This is now in Issue #419

File: linetools/isgm/emsystem.py (hunk 10 above)
Comment (tiffanyhsyu) [reply linetools/isgm/emsystem.py]: Just checked that this works! There is also a test for it now.

File: linetools/tests/test_io.py (hunk 12 above)
Comment (tiffanyhsyu) [reply linetools/tests/test_io.py]: Done

File: linetools/isgm/tests/test_init_emsys.py (hunk 11 above)
Comment (tiffanyhsyu) [reply linetools/isgm/tests/test_init_emsys.py]: I've added a test for creating an EmSystem object

**Comment by profxj** [comment]
Hey @tiffanyhsyu -- One test is still failing:

```

def test_emline_from_alis():
    J0018_emlines = lio.emlines_from_alis_output(data_path('spec1d_J0018p2345_KASTb_coadd.mod.out'))

    assert len(J0018_emlines) == 8
    np.testing.assert_allclose(J0018_emlines[0].z, 0.01540425)
    np.testing.assert_allclose(J0018_emlines[1].wrest.value, 4341.684) #Hgamma
```

That is because the file is not in the repo.  And you may need to edit one
other thing too. But add that file in first.  Thanks

**Comment by profxj** [comment]
The test, by the way, is  

tests/test_io.py

**Comment by tiffanyhsyu** [comment]
@profxj I've added a 'files' folder under 'test' and uploaded the ALIS .mod.out file there -- is that ok?

**Comment by profxj** [comment]
merging!

**Pull request merged**